common/tools/dev-tool/src/util/samples/processor.ts

```diff
@@ -0,0 +1,405 @@
+    // above and the CommonJS transforms (see transforms.ts).
+    const jsModuleText = convert(sourceText, {
+      fileName: source,
+      transformers: {
+        before: [sourceProcessor],
+        after: [toCommonJs],
+      },
+    });
+
+    // Check the imports for any relative module imports (these could be util files), and compute a relative path to the
+    // module from the source directory
+    for (const relativeModule of accumulator.importedModules
+      .filter(isRelativePath)
+      .map((modulePath) =>
+        path.normalize(path.join(path.dirname(relativeSourcePath), modulePath))
+      )) {
+      importedRelativeModules.add(relativeModule);
+    }
+
+    return {
+      filePath: source,
+      relativeSourcePath,
+      text: sourceText,
+      jsModuleText,
+      summary,
+      importedModules: accumulator.importedModules.filter(isDependency),
+      usedEnvironmentVariables: accumulator.usedEnvironmentVariables,
+      azSdkTags,
+    };
+  });
+
+  return Promise.all(jobs).then((results) => {
+    // Only fail once at the end, so that we don't drown you with tons of red messages.
+    // Think of this whole *then* continuation as a validation pass.
+
+    if (hadUnsupportedSyntax) {
+      fail(
+        "Samples must support the latest Node LTS well. See the errors above for more information."
+      );
+    }
+
+    for (const result of results) {
+      if (
+        result.summary === undefined &&
+        !importedRelativeModules.has(result.relativeSourcePath.replace(/\.ts$/, "")) &&
+        !result.azSdkTags.util
+      ) {
+        fail(
+          `${result.relativeSourcePath} does not include an @summary tag, is not imported by any other module, and is not marked as a util (using @azsdk-util true).`
+        );
+      }
+    }
+
+    return results;
+  });
+}
+
+function isImportOrStaticRequire(node: ts.Node): node is ts.ImportDeclaration | ts.CallExpression {
+  return (
+    ts.isImportDeclaration(node) ||
+    (ts.isCallExpression(node) &&
+      ((ts.isIdentifier(node.expression) && node.expression.text === "require") ||
+        node.expression.kind === ts.SyntaxKind.ImportKeyword) &&
+      ts.isStringLiteralLike(node.arguments[0]))
+  );
+}
+
+/**
+ * Detects usage of environment variables.
+ *
+ * This can _only_ work with property access expressions where the left hand side is a node that has the exact text
+ *
+ * "process.env"
+ *
+ * For example, it won't work with process["env"], and it won't work if process is bound to another name. It will _only_
+ * work with:
+ *
+ * - `process.env.NAME`
+ * - `process.env["NAME"]`
+ *
+ * @param node - the node to test
+ * @param sourceFile - the source file where the node appears
+ * @returns true if the node appears to be a process.env access.
+ */
+function isProcessEnvAccess(
+  node: ts.Node,
+  sourceFile?: ts.SourceFile
+): node is ts.PropertyAccessExpression | ts.ElementAccessExpression {
```

Comment:
Technically-incorrect use of type predicates here - a user-defined type guard should return true for *any* value matching its predicate, not some of them.

It is not strictly necessary to address this.

Reply:
Now that you've mentioned this, I understand why that's the case, because the compiler might refine the type to exclude the types in the guard if it can falsify it. I also don't think the guard is actually doing any work for us anymore (this was used previously, and we did rely on the type guard, but I don't think it's relevant anymore). I'll go back and change this to a simple boolean.